File: tests/python/tests-python-pytest-requests-selenium/tokens_css.py
"""tokens.css resolver — mirrors Java TokensCss."""

from __future__ import annotations

import re
from pathlib import Path

_ROOT_BLOCK = re.compile(r":root\s*\{([^}]+)\}", re.DOTALL)
_TOKEN = re.compile(r"(--[\w-]+)\s*:\s*([^;]+);")

_MODULE_ROOT = Path(__file__).resolve().parent
_APP_ROOT = (_MODULE_ROOT / ".." / ".." / "..").resolve()


def default_tokens_path() -> Path:
    return resolve_from_app_root(_APP_ROOT)


def resolve_from_app_root(app_root: Path) -> Path:
    return first_existing(*tokens_css_candidates(app_root))


def first_existing(*candidates: Path) -> Path:
    fallback = candidates[-1].resolve()
    for candidate in candidates:
        abs_path = candidate.resolve()
        if abs_path.exists():
            return abs_path
        fallback = abs_path
    return fallback


def parse_root_tokens(css_file: Path) -> dict[str, str]:
    css = css_file.read_text(encoding="utf-8")
    match = _ROOT_BLOCK.search(css)
    if not match:
        raise ValueError(f":root block not found in {css_file}")
    tokens: dict[str, str] = {}
    for token, value in _TOKEN.findall(match.group(1)):
        tokens[token] = value.strip()
    return tokens


def tokens_css_candidates(app_root: Path) -> list[Path]:
    candidates = [hub_tokens(app_root)]
    _append_vendor_tokens(app_root / "frontend", candidates)
    return candidates


def hub_tokens(app_root: Path) -> Path:
    return app_root / "frontend" / "_shared" / "frontend-javascript-app" / "css" / "tokens.css"


def _append_vendor_tokens(frontend_root: Path, out: list[Path]) -> None:
    if not frontend_root.is_dir():
        return
    for lang in sorted(p for p in frontend_root.iterdir() if p.is_dir()):
        if not _is_product_language_dir(lang):
            continue
        for cell in sorted(p for p in lang.iterdir() if p.is_dir()):
            if cell.name.startswith("."):
                continue
            out.append(cell / "vendor" / "ds" / "css" / "tokens.css")
            out.append(cell / "vendor" / "frontend-javascript-app" / "css" / "tokens.css")


def _is_product_language_dir(path: Path) -> bool:
    name = path.name
    return not name.startswith(".") and not name.startswith("_") and name not in {"scripts", "node_modules"}

File: tests/typescript/tests-typescript-api_request-playwright/src/helpers/tokens-css.ts
// @ts-nocheck
import fs from 'fs';
import path from 'path';

const MODULE_ROOT = path.resolve(__dirname, '../..');
const APP_ROOT = path.resolve(MODULE_ROOT, '../../..');
const ROOT_BLOCK = /:root\s*\{([^}]+)\}/s;
const TOKEN = /(--[\w-]+)\s*:\s*([^;]+);/g;

export function defaultTokensPath() {
  return resolveFromAppRoot(APP_ROOT);
}

export function resolveFromAppRoot(appRoot: string) {
  return firstExisting(tokensCssCandidates(appRoot));
}

export function firstExisting(candidates: string[]) {
  let fallback = path.resolve(candidates[candidates.length - 1]);
  for (const candidate of candidates) {
    const abs = path.resolve(candidate);
    if (fs.existsSync(abs)) {
      return abs;
    }
    fallback = abs;
  }
  return fallback;
}

export function parseRootTokens(cssFile: string) {
  const css = fs.readFileSync(cssFile, 'utf8');
  const match = css.match(ROOT_BLOCK);
  if (!match) {
    throw new Error(`:root block not found in ${cssFile}`);
  }
  const tokens: Record<string, string> = {};
  for (const tokenMatch of match[1].matchAll(TOKEN)) {
    tokens[tokenMatch[1]] = tokenMatch[2].trim();
  }
  return tokens;
}

function tokensCssCandidates(appRoot: string) {
  const candidates = [hubTokens(appRoot)];
  appendVendorTokens(path.join(appRoot, 'frontend'), candidates);
  return candidates;
}

function hubTokens(appRoot: string) {
  return path.join(appRoot, 'frontend', '_shared', 'frontend-javascript-app', 'css', 'tokens.css');
}

function appendVendorTokens(frontendRoot: string, out: string[]) {
  if (!fs.existsSync(frontendRoot)) {
    return;
  }
  const langs = fs
    .readdirSync(frontendRoot, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .sort();
  for (const lang of langs) {
    if (!isProductLanguageDir(lang)) {
      continue;
    }
    const langDir = path.join(frontendRoot, lang);
    const cells = fs
      .readdirSync(langDir, { withFileTypes: true })
      .filter((entry) => entry.isDirectory())
      .map((entry) => entry.name)
      .sort();
    for (const cell of cells) {
      if (cell.startsWith('.')) {
        continue;
      }
      const cellDir = path.join(langDir, cell);
      out.push(path.join(cellDir, 'vendor', 'ds', 'css', 'tokens.css'));
      out.push(path.join(cellDir, 'vendor', 'frontend-javascript-app', 'css', 'tokens.css'));
    }
  }
}

function isProductLanguageDir(name: string) {
  return !name.startsWith('.') && !name.startsWith('_') && name !== 'scripts' && name !== 'node_modules';
}
